### PR TITLE
Issue #664: modify dataset config structure

### LINKED ISSUE #664: Separate data config for different splits
## Subject of the issue

When predict with config labeled=true, without providing label files, error raised:
WARNING:root:sample_label is not used in predict. It is True if and only if mode == 'train'.
WARNING:root:Using customized configuration.The code might break if the config doesn't match the saved model.
2021-02-02 19:58:43.570942: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory
2021-02-02 19:58:43.570972: W tensorflow/stream_executor/cuda/cuda_driver.cc:312] failed call to cuInit: UNKNOWN ERROR (303)
2021-02-02 19:58:43.570986: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (precision7750-yhu): /proc/driver/nvidia/version does not exist
Traceback (most recent call last):
  File "demos/paired_ct_lung/demo_predict.py", line 59, in <module>
    config_path=config_path,
  File "/home/yipenghu/git/DeepReg/deepreg/predict.py", line 238, in predict
    repeat=False,
  File "/home/yipenghu/git/DeepReg/deepreg/util.py", line 42, in build_dataset
    data_loader = get_data_loader(dataset_config, mode)
  File "/home/yipenghu/git/DeepReg/deepreg/dataset/load.py", line 47, in get_data_loader
    config=data_loader_config, default_args=default_args
  File "/home/yipenghu/git/DeepReg/deepreg/registry.py", line 262, in build_data_loader
    category=DATA_LOADER_CLASS, config=config, default_args=default_args
  File "/home/yipenghu/git/DeepReg/deepreg/registry.py", line 141, in build_from_config
    return cls(**args)
  File "/home/yipenghu/git/DeepReg/deepreg/dataset/loader/paired_loader.py", line 71, in __init__
    dir_paths=data_dir_paths, name="moving_labels", grouped=False
  File "/home/yipenghu/git/DeepReg/deepreg/dataset/loader/nifti_loader.py", line 40, in __init__
    self.set_data_structure()
  File "/home/yipenghu/git/DeepReg/deepreg/dataset/loader/nifti_loader.py", line 65, in set_data_structure
    ), f"directory {named_dir_path} does not exist"
AssertionError: directory demos/paired_ct_lung/dataset/test/moving_labels does not exist

## Your environment

- DeepReg version bd3202ace94151d657c3daf9c49ce06563fdc556

  

- OS: Ubuntu 20.04

- Python Version: 3.7

- TensorFlow

  - TensorFlow Version 2.3

## Steps to reproduce

remove fixed_labels and moving_labels in dataset/test folder of the demo, e.g. paired_ct_lung

run `python demos/paired_ct_lung/demo_predict.py`

## Expected behaviour

It should allow prediction without any labels

## Actual behaviour

1- It only allows prediction when the config key changed to "false" - this is a confusing behaviour as the labeled key should only control the training.

2- When debugging, the docstring isn't clear:
L163 in network.py:
it is not clear when labels are not provided, if the labels should be "None" or else. 

3- double check: 
The availability of the labels should not affect the model output.

@mathpluscode can you have a look.

### PR DESCRIPTION
**Pull request by mathpluscode**
# Description

Fixes #664

## Type of change

What types of changes does your code introduce to DeepReg?

_Please check the boxes that apply after submitting the pull request._

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Documentation Update (fix or improvement on the documentation)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Other (if none of the other choices apply)

## Checklist

_Please check the boxes that apply after submitting the pull request._

_If you're unsure about any of them, don't hesitate to ask. We're here to help! This is
simply a reminder of what we are going to look for before merging your code._

- [x] I have
      [installed pre-commit](https://deepreg.readthedocs.io/en/latest/contributing/setup.html)
      using `pre-commit install` and formatted all changed files. If you are not
      certain, run `pre-commit run --all-files`.
- [x] My commits' message styles matches
      [our requested structure](https://deepreg.readthedocs.io/en/latest/contributing/commit.html),
      e.g. `Issue #<issue number>: detailed message`.
- [x] I have updated the
      [change log file](https://github.com/DeepRegNet/DeepReg/blob/main/CHANGELOG.md)
      regarding my changes.
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by mathpluscode** [comment]
@YipengHu @NMontanaBrown @kate-sann5100 In this PR, I've modified the config such that different split defines its own `format` and `labeled`. I also renamed `mode` to `split` to be clear, so that it's for data, not for the model.

Regarding
- train with labels, and evaluate without labels.
- or train without labels, and evaluate with labels.
I will open a new issue and handle it in a separate PR, to prevent from modifying 100 files at a time.

Please kindly review this PR if you have time ^^

**Comment by NMontanaBrown** [comment]
I don't really get the point of this set of changes.

WRT changing the order of train, valid, test, it just seems a style refactor of the config file, which adds new changes and doesn't necessarily improve UX. I don't understand the "bugfix" of the PR. Therefore it seeems like a bit unnecessary changes to the code. @YipengHu @mathpluscode

**Comment by YipengHu** [comment]
> I don't really get the point of this set of changes.
> 
> WRT changing the order of train, valid, test, it just seems a style refactor of the config file, which adds new changes and doesn't necessarily improve UX. I don't understand the "bugfix" of the PR. Therefore it seeems like a bit unnecessary changes to the code. @YipengHu @mathpluscode 

It is the first step to fixing the bug in #664, as commented above and in #664. You need to separate these config fields to enable the separation. Which specific changes you think is unnecessary and, in that case, please suggest alternative to fix #664?

**Comment by mathpluscode** [comment]
> I don't really get the point of this set of changes.
> 
> WRT changing the order of train, valid, test, it just seems a style refactor of the config file, which adds new changes and doesn't necessarily improve UX. I don't understand the "bugfix" of the PR. Therefore it seeems like a bit unnecessary changes to the code. @YipengHu @mathpluscode

@NMontanaBrown 

First, without this PR, we will not be able to have different "label" conditions for different splits, which means that we can not train without labels, then evaluate with labels. Or train with labels, then evaluate without labels.

Second, it is a bug as DDF models should not require labels as input, which means it is a bug when we do inference with DDF models, but deepreg raised an error if the labels do not exist. If you insist, we can consider it as refactoring or whatever, but it is minor details.

Third, this PR not only defines the label per split but also the format, which means that we have more possibilities to combine datasets.

Therefore, I would not consider this PR unnecessary. Are you convinced now? :)

**Comment by NMontanaBrown** [comment]
Is this ready for re-review? @mathpluscode

**Comment by mathpluscode** [comment]
> Is this ready for re-review? @mathpluscode 

@NMontanaBrown Yes! Please! I was just synchronizing the branch.

**Review by NMontanaBrown — APPROVED** [review]

**Review by YipengHu — APPROVED** [review]

File: deepreg/config/parser.py [hunk 1]
```diff
@@ -80,15 +80,12 @@ def config_sanity_check(config: dict) -> dict:
...
-
     # back compatibility support
     config = parse_v011(config)
 
     # check model
     if config["train"]["method"] == "conditional":
-        if data_config["labeled"] is False:  # unlabeled
+        if config["dataset"]["train"]["labeled"] is False:  # unlabeled
             raise ValueError(
                 "For conditional model, data have to be labeled, got unlabeled data."
```
Comment (YipengHu) [inline deepreg/config/parser.py:90]: just a note here: for conditional seg. only moving images "have to" be labelled for val and test splits.

Comment (mathpluscode) [reply deepreg/config/parser.py:90]: Good point! But that's yet another level of complexity :(

**Pull request merged**